_pytest/python_api.py

```diff
@@ -6,10 +6,271 @@
+        to be on either side of the `==` operator.  The reason for this has to 
+        do with how python decides whether to call `a.__eq__()` or `b.__eq__()` 
+        when it encounters `a == b`.  
+        
+        If `a` and `b` are not related by inheritance, `a` gets priority.  So 
+        as long as `a.__eq__` is defined, it will be called.  Because most 
+        implementations of `a.__eq__` end up calling `b.__eq__`, this detail 
+        usually doesn't matter.  However, `numpy.ndarray.__eq__` raises an 
+        error complaining that "the truth value of an array with more than 
+        one element is ambiguous. Use a.any() or a.all()" when compared with a 
+        custom class, so `b.__eq__` never gets called.
+
+        The trick is that the priority rules change if `a` and `b` are related 
+        by inheritance.  Specifically, `b.__eq__` gets priority if `b` is a 
+        subclass of `a`.  So we can guarantee that `ApproxNumpy.__eq__` gets 
+        called by inheriting from `numpy.ndarray`.
+        """
+
+        def __new__(cls, expected, rel=None, abs=None):
+            """
+            Numpy uses __new__ (rather than __init__) to initialize objects. 
+            
+            The `expected` argument must be a numpy array.  This should be 
+            ensured by the approx() delegator function.
+            """
+            assert isinstance(expected, np.ndarray)
+            obj = super(ApproxNumpy, cls).__new__(cls, expected.shape)
+            obj.__init__(expected, rel, abs)
+            return obj
+
+        def __repr__(self):
+            # It might be nice to rewrite this function to account for the 
+            # shape of the array...
+            return '[' + ApproxBase.__repr__(self) + ']'
+
+        def __eq__(self, actual):
+            try:
+                actual = np.array(actual)
+            except:
+                raise ValueError("cannot cast '{0}' to numpy.ndarray".format(actual))
+
+            if actual.shape != self.expected.shape:
+                return False
+
+            return ApproxBase.__eq__(self, actual)
+
+        def _yield_expected(self):
+            for x in self.expected:
+                yield x
+
+        def _yield_comparisons(self, actual):
+            # We can be sure that `actual` is a numpy array, because it's 
+            # casted in `__eq__` before being passed to `ApproxBase.__eq__`, 
+            # which is the only method that calls this one.
+            for i in np.ndindex(self.expected.shape):
+                yield actual[i], self.expected[i]
+
+
+except ImportError:
+    np = None
+
+class ApproxMapping(ApproxBase):
+    """
+    Perform approximate comparisons for mappings where the values are numbers 
+    (the keys can be anything).
+    """
+
+    def __repr__(self):
+        item = lambda k, v: "'{0}': {1}".format(k, self._approx_scalar(v))
+        return '{' + ', '.join(item(k,v) for k,v in self.expected.items()) + '}'
```

Comment:
The approx `__repr__` is still distinguishable from the dict/list/tuple `__repr__` because tolerances are included with each value.  In other words, `repr(approx(1.0))` will look like `{'a': 1.0 ± 1e-6}` rather than `{'a': 1.0}`.

Also, the reason I tried to mimic the format of the built-in data structures in the first place is that when I get assertion failures, I find it easier to spot the important differences if the actual and expected values are formatted the same.  If they're formatted differently (i.e. one has braces and the other doesn't), my eye immediately goes to those differences first.  So I think the braces improve clarity more than they might detract from it.